builtin/credential/appgroup/user_id.go

```diff
@@ -0,0 +1,372 @@
+
+		resp.TokenTTL = group.TokenTTL
+		resp.TokenMaxTTL = group.TokenMaxTTL
+		resp.Wrapped = group.Wrapped
+	case selectorTypeGeneric:
+		generic, err := b.genericEntry(s, selectorValue)
+		if err != nil {
+			return nil, err
+		}
+		if generic == nil {
+			return nil, fmt.Errorf("generic credential referred by the user ID does not exist")
+		}
+		for _, groupName := range generic.Groups {
+			group, err := b.groupEntry(s, groupName)
+			if err != nil {
+				return nil, err
+			}
+			groupPolicies, err := b.fetchPolicies(s, group.Apps)
+			if err != nil {
+				return nil, err
+			}
+			// Append the union of all the policies from all the apps on the group
+			resp.Policies = append(resp.Policies, groupPolicies...)
+			// Append the additional policies set on the group
+			resp.Policies = append(resp.Policies, group.AdditionalPolicies...)
+		}
+
+		for _, appName := range generic.Apps {
+			app, err := b.appEntry(s, appName)
+			if err != nil {
+				return nil, err
+			}
+			// Append the policies set on the app
+			resp.Policies = append(resp.Policies, app.Policies...)
+		}
+
+		// Append the additonal policies set on the generic entry
+		resp.Policies = append(resp.Policies, generic.AdditionalPolicies...)
+
+		resp.TokenTTL = generic.TokenTTL
+		resp.TokenMaxTTL = generic.TokenMaxTTL
+		resp.Wrapped = generic.Wrapped
+	default:
+		return nil, fmt.Errorf("unknown selector type")
+	}
+
+	// Cap the token_ttl and token_max_ttl values.
+	var err error
+	resp.TokenTTL, resp.TokenMaxTTL, err = b.SanitizeTTL(resp.TokenTTL, resp.TokenMaxTTL)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.Policies = policyutil.SanitizePolicies(resp.Policies)
+
+	// Even though wrapped is unrelated to the token_ttl and token_max_ttl
+	// values, since it is issued out of the backend, it should respect the
+	// backend's boundaries.
+	if resp.Wrapped > b.System().MaxLeaseTTL() {
+		resp.Wrapped = b.System().MaxLeaseTTL()
+	}
+
+	return resp, nil
+}
+
+// userIDEntryValid is used to determine if the given user ID is a valid one.
+// The UserID is looked to be present only under the sub-view of the selector.
+// This ensures that the UserIDs that are reused between selector types, the
+// correct one is referred to. If the UserIDs are always generated by the
+// backend, then there will be no collision between the UserIDs from different
+// types. But, if same specific UserIDs are assigned across different selector
+// types, then it should be supported.
+func (b *backend) userIDEntryValid(s logical.Storage, selectorType, selectorValue, userID string) (bool, error) {
+	// If there is no lock indexed by the UserID in the map of locks,
+	// it implies that there is no storage entry corresponding to the
+	// UserID. It is either that the ID is an invalid one or that the
+	// UserID is expired. Fail the validation.
+	b.userIDLocksMapLock.RLock()
+	lock := b.userIDLocksMap[userID]
+	if lock == nil {
+		return false, nil
```

Comment:
You'll return here without releasing the lock on the map. You could defer the `RUnlock` but I think here it's safe to simply `RUnlock` the map immediately after fetching `lock`.